econml/validate/drtester.py

```diff
@@ -0,0 +1,796 @@
+        rsq = round(self.cal_r_squared[np.where(self.tmts == tmt)[0][0] - 1], 3)
+        df['95_err'] = 1.96 * df['se_gate']
+        fig = df.plot(
+            kind='scatter',
+            x='g_cate',
+            y='gate',
+            yerr='95_err',
+            xlabel = 'Group Mean CATE',
+            ylabel = 'GATE',
+            title=f"Treatment = {tmt}, Calibration R^2 = {rsq}"
+        )
+
+        return fig
+
+    def evaluate_blp(
+        self,
+        reg_cate=None,
+        Zval: np.array = None,
+        Ztrain: np.array = None
+    ):
+        """
+        Implements the best linear predictor (BLP) test as in [Chernozhukov2022]. `fit_nusiance' method must already
+        be implemented.
+
+        Parameters
+        ----------
+        reg_cate: estimator, default ``None''
+            CATE model. Must be able to implement `fit' and `predict' methods. If not specified, then fit_cate() must
+            already have been implemented
+        Zval: (n_val x k) matrix, default ``None''
+            Validation sample features for CATE model. If not specified, then `fit_cate' method must already have been
+            implemented
+        Ztrain: (n_train x k) matrix, default ``None''
+            Training sample features for CATE model. If specified, then CATE is fitted on training sample and applied
+            to Zval. If specified, then Xtrain, Dtrain, Ytrain must have been specified in `fit_nuisance' method (and
+            vice-versa)
+
+        Returns
+        -------
+        self, with added dataframe blp_res showing the results of the BLP test
+
+        """
+
+        if self.dr_val is None:
+            raise Exception("Must fit nuisances before evaluating")
+
+        # if CATE is given explicitly or has not been fitted at all previously, fit it now
+        if (self.cate_preds_val is None) or (reg_cate is not None):
+            if (Zval is None) or (reg_cate is None):  # need at least Zval and a CATE estimator to fit
+                raise Exception('CATE not yet fitted - must provide Zval and CATE estimator')
+            self.fit_cate(reg_cate, Zval, Ztrain)
+
+        if self.n_treat == 1:  # binary treatment
+            reg = OLS(self.dr_val, add_constant(self.cate_preds_val)).fit()
+            params = [reg.params[1]]
+            errs = [reg.bse[1]]
+            pvals = [reg.pvalues[1]]
+        else:  # categorical treatment
+            params = []
+            errs = []
+            pvals = []
+            for k in range(self.n_treat):  # run a separate regression for each
+                reg = OLS(self.dr_val[:, k], add_constant(self.cate_preds_val[:, k])).fit(cov_type = 'HC1')
+                params.append(reg.params[1])
+                errs.append(reg.bse[1])
+                pvals.append(reg.pvalues[1])
+
+        self.blp_res = pd.DataFrame(
+            {'treatment': self.tmts[1:], 'blp_est': params, 'blp_se': errs, 'blp_pval': pvals}
+        ).round(3)
+
+        return self
```

Comment:
Similarly maybe here this should be returning a instance a class `BLPEvaluationResults`